tree/dataframe/src/RSqliteDS.cxx

```diff
@@ -13,19 +13,304 @@
+
+constexpr char const *gSQliteVfsName = "ROOT-Davix-readonly";
+
+#ifdef R__HAS_DAVIX
+
+/**
+ * The "derived class" from sqlite3_file that gets passed to the file operation callbacks.
+ */
+struct VfsRootFile {
+   VfsRootFile() : pos(&c) {}
+   sqlite3_file pFile;
+   DAVIX_FD *fd;
+   uint64_t size;
+   Davix::Context c;
+   Davix::DavPosix pos;
+};
+
+////////////////////////////////////////////////////////////////////////////
+/// Releases the resources associated to a file opened with davix
+static int VfsRdOnlyClose(sqlite3_file *pFile)
+{
+   Davix::DavixError *err = nullptr;
+   VfsRootFile *p = reinterpret_cast<VfsRootFile *>(pFile);
+   if (p->pos.close(p->fd, &err) == -1) {
+      p->~VfsRootFile();
+      return SQLITE_IOERR_CLOSE;
+   }
+   p->~VfsRootFile();
+   return SQLITE_OK;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// Issues an HTTP range request to read a chunk from a remote file
+static int VfsRdOnlyRead(sqlite3_file *pFile, void *zBuf, int count, sqlite_int64 offset)
+{
+   Davix::DavixError *err = nullptr;
+   VfsRootFile *p = reinterpret_cast<VfsRootFile *>(pFile);
+   if (p->pos.pread(p->fd, zBuf, count, offset, &err) == -1) {
+      return SQLITE_IOERR;
+   }
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyWrite(sqlite3_file * /*pFile*/, const void * /*zBuf*/, int /*iAmt*/, sqlite_int64 /*iOfst*/)
+{
+   return SQLITE_OPEN_READONLY;
+}
+
+static int VfsRdOnlyTruncate(sqlite3_file * /*pFile*/, sqlite_int64 /*size*/)
+{
+   return SQLITE_OPEN_READONLY;
+}
+
+static int VfsRdOnlySync(sqlite3_file * /*pFile*/, int /*flags*/)
+{
+   return SQLITE_OK;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// Returns the cached file size
+static int VfsRdOnlyFileSize(sqlite3_file *pFile, sqlite_int64 *pSize)
+{
+   VfsRootFile *p = reinterpret_cast<VfsRootFile *>(pFile);
+   *pSize = p->size;
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyLock(sqlite3_file * /*pFile*/, int /*level*/)
+{
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyUnlock(sqlite3_file * /*pFile*/, int /*level*/)
+{
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyCheckReservedLock(sqlite3_file * /*pFile*/, int *pResOut)
+{
+   *pResOut = 0;
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyFileControl(sqlite3_file * /*p*/, int /*op*/, void * /*pArg*/)
+{
+   return SQLITE_NOTFOUND;
+}
+
+static int VfsRdOnlySectorSize(sqlite3_file * /*pFile*/)
+{
+   return SQLITE_OPEN_READONLY;
+}
+
+static int VfsRdOnlyDeviceCharacteristics(sqlite3_file * /*pFile*/)
+{
+   return SQLITE_OPEN_READONLY;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// Fills a new VfsRootFile struct enclosing a Davix file
+static int VfsRdOnlyOpen(sqlite3_vfs * /*vfs*/, const char *zName, sqlite3_file *pFile, int flags, int * /*pOutFlags*/)
+{
+   VfsRootFile *p = new (pFile) VfsRootFile();
+   p->pFile.pMethods = nullptr;
+
+   static const sqlite3_io_methods io_methods = {
```

Comment:
please comment :) I can guess what's going on because I once glimpsed into the abyss of the cpython C api :P